File: ui/cypress/tests/adapter/machineDataSimulator.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { AdapterUtils } from '../../support/utils/AdapterUtils';
import { SpecificAdapterBuilder } from '../../support/builder/SpecificAdapterBuilder';

describe('Test Random Data Simulator Stream Adapter', () => {
  before('Setup Test', () => {
    cy.initStreamPipesTest();
  });

  it('Perform Test', () => {
    const adapterInput = SpecificAdapterBuilder
      .create('Machine_Data_Simulator')
      .setName('Machine Data Simulator Test')
      .addInput('input', 'wait-time-ms', '1000')
      .setTimestampProperty('timestamp')
      .setStoreInDataLake()
      .build();

    AdapterUtils.testSpecificStreamAdapter(adapterInput);

    // const adapterInput1 = SpecificAdapterBuilder
    //   .create('Machine_Data_Simulator')
    //   .setName('Machine Data Simulator Test 2')
    //   .addInput('input', 'wait-time-ms', '1000')
    //   .addInput('radio', 'selected-simulator-option', 'pressure')
    //   .setTimestampProperty('timestamp')
    //   .setStoreInDataLake()
    //   .build();
    //
    // AdapterUtils.testSpecificStreamAdapter(adapterInput1);
    // AdapterUtils.deleteAdapter();
  });

});

File: streampipes-extensions/streampipes-connect-adapters-iiot/src/main/java/org/apache/streampipes/connect/iiot/adapters/simulator/machine/MachineDataSimulator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package org.apache.streampipes.connect.iiot.adapters.simulator.machine;

import org.apache.streampipes.connect.iiot.adapters.simulator.machine.event.EventSimulator;
import org.apache.streampipes.extensions.api.connect.IEventCollector;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;

public class MachineDataSimulator implements Runnable {

  private final IEventCollector collector;

  private final Integer waitTimeMs;
  private final Integer numberOfSensors;
  private final EventSimulator simulator;

  private Boolean running;

  private static final Logger LOG = LoggerFactory.getLogger(MachineDataSimulator.class);

  public MachineDataSimulator(EventSimulator simulator,
                              IEventCollector collector,
                              Integer waitTimeMs,
                              int numberOfSensors) {
    this.simulator = simulator;
    this.collector = collector;
    this.waitTimeMs = waitTimeMs;
    this.running = true;
    this.numberOfSensors = numberOfSensors;
  }

  @Override
  public void run() {
    this.running = true;
    long startTimeMs = System.currentTimeMillis();

    while (running) {
      long currentTimeMs = System.currentTimeMillis();
      long timeDeltaMs = currentTimeMs - startTimeMs;

      Integer simulationPhase = null;

      if (timeDeltaMs > 0 && timeDeltaMs <= 30000) {
        simulationPhase = 0;
      } else if (timeDeltaMs > 30000 && timeDeltaMs <= 60000) {
        simulationPhase = 1;
      } else {
        startTimeMs = currentTimeMs; // reset
      }

      if (simulationPhase != null) {
        long timestamp = System.currentTimeMillis();

        for (int sensorIndex = 1; sensorIndex <= numberOfSensors; sensorIndex++) {
          var event = simulator.buildEvent(simulationPhase, sensorIndex, timestamp);

          if (!event.isEmpty()) {
            collector.collect(event);
          }
        }
      }

      try {
        TimeUnit.MILLISECONDS.sleep(waitTimeMs);
      } catch (InterruptedException e) {
        LOG.error("Machine simulator thread interrupted", e);
        Thread.currentThread().interrupt();
        return;
      }
    }
  }



  public void setRunning(Boolean running) {
    this.running = running;
  }
}
